text/3446-store.md

```diff
@@ -0,0 +1,998 @@
+
+The _Safety_ section of the [`Allocator`](https://doc.rust-lang.org/nightly/std/alloc/trait.Allocator.html#safety)
+documentation notes that a `Clone` of an `Allocator` must be interchangeable with the original, and that all allocated
+pointers must remain until the last of the clones or copies of the allocator is dropped.
+
+The standard library then proceed to require `A: Allocator + Clone` to clone a `Box` or a `Vec`, when arguably it is not
+necessary to have an interchangeable allocator, and instead _semantically_ an independent allocator is required.
+
+This RFC, instead, favors using the `Default` bound for the `Clone` implementation of `StoreBox`:
+
+1.  It matches the desired semantics better -- a brand new store is required, not an interchangeable one.
+2.  A cloneable `InlineStore` cannot match the semantics of `Clone` required of Allocators.
+
+`Default` does have the issue that it may not mesh well with `dyn Store` or `dyn Allocator`, and while `Clone` can
+reasonably be implemented for `&dyn Store`, or `Rc<dyn Store>`, such is not the case for `Default`.
+
+This leaves 4 possibilities:
+
+-   Use `Clone` despite the poor semantics match.
+-   Use `Default` despite it being at odds with `dyn Store` use.
+-   Add a new `SpawningStore` trait to create an independent instance, though mixing several non-empty traits in a `dyn`
+    context is not supported yet.
+-   Add a method to `Store` to create an independent instance, fixing the semantics of `Clone`. Possibly a faillible
+    one.
+
+Note that technically switching from the `Clone` bound to another bound for `Box` and `Vec` is a breaking change,
+however since `Allocator` is an unstable API it is still early enough to effect such change.
+
+
+##  (Minor) To `Clone` or to share?
+
+As mentioned above, whenever an `Allocator` also implements the `Clone` trait, the clone or copy of the `Allocator` must
+fulfill specific requirements. In particular, all clones or copies of a given allocator behave as a single allocator
+sharing the backing memory and metadata. While the `Clone` trait does fit _creating_ a new clone/copy of an allocator,
+it is insufficient however to _query_ whether another instance is a clone/copy of a given allocator.
+
+The standard library runs headlong into this insufficience, and while `LinkedList::split_off` is implemented for any
+`Allocator` which also implements `Clone`, `LinkedList::append` is only implemented for `Global`.
+
+There are at least 2 possibilities, here:
+
+-   Add a requirement on `PartialEq` implementation for `Allocator` and `Store` that comparing equal means that they are
+    clones or copies of each others.
+-   Add a separate `SharingStore` trait -- see future possibilities.
+
+It should be noted that `dyn` usage of `Allocator` and `Store` suffers from the requirement of using unrelated traits as
+it is not possible to have a `dyn Allocator + Clone + PartialEq` trait today, though those traits can be implemented for
```

Comment:
Multiple trait vtables are coming "soon," along with vtable upcasting.

I am now thinking we might potentially want a separate `trait ErasedStore` from the `trait Store` for the purpose of dynamic storages, to shim `ErasedHandle` in and potentially to provide `is_sharing_with` and `clone` functionality. Making `Store` deliberately *not* object safe initially might thus be desirable, until we figure out how dynamic storages "should" work.

Reply:
No opinion \o/